File: code/F-MPI/traffic.f90
program traffic

  use mpi

  use trafficlib

  implicit none

! Set the size of the road

  integer :: ncell = 5120000
  integer :: maxiter, printfreq

  integer rank, size, ierr, nlocal, rankup, rankdown
  integer, dimension(MPI_STATUS_SIZE) :: status

  integer :: i, iter, nmove, nmovelocal, ncars, ncars_local
  real    :: density

  integer, allocatable, dimension(:) :: newroad, oldroad

  double precision :: tstart, tstop

  maxiter = 200
  printfreq = maxiter/10

! Set target density of cars

  density = 0.52

! Start message passing system

  call MPI_Init(ierr)

! Compute size and rank

  call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  ncell = ncell * size
  nlocal = ncell/size

  if (rank == 0) then

     write(*,*) 'Running message-passing traffic model'
     write(*,*)
     write(*,*) 'Length of road is ', ncell
     write(*,*) 'Number of iterations is ', maxiter
     write(*,*) 'Target density of cars is ', density
     write(*,*) 'Running on ', size, ' processes'

  end if

! Allocate arrays

  allocate(newroad(0:nlocal+1))
  allocate(oldroad(0:nlocal+1))

  do i = 1, nlocal
     oldroad(i) = 0
     newroad(i) = 0
  end do

  if (rank == 0) then
    ! Initialise road accordingly using random number generator
    write(*,*) 'Initialising ...'
  end if

  call rinit(5743)
  ncars_local = 0
  do i = 0, rank
    ncars_local = initroad(oldroad(1), nlocal, density)
  end do
  call MPI_Reduce(ncars_local, ncars, 1, MPI_INTEGER, MPI_SUM, &
      0, MPI_COMM_WORLD, ierr)

  if (rank == 0) then
     write(*,*) '... done'

     write(*,*) 'Actual density of cars is ', float(ncars)/float(ncell)
     write(*,*) 'Scattering data ...'

  end if

  if (rank == 0) then

     write(*,*) '... done'
     write(*,*)

  end if

! Compute neighbours

  rankup   = mod(rank + 1, size)
  rankdown = mod(rank + size - 1, size)

  tstart = MPI_Wtime()

  do iter = 1, maxiter

! Implement halo swaps which now includes boundary conditions

     call MPI_Sendrecv(oldroad(nlocal), 1, MPI_INTEGER, rankup, 1,   &
                       oldroad(0),      1, MPI_INTEGER, rankdown, 1, &
                       MPI_COMM_WORLD, status, ierr);

     call MPI_Sendrecv(oldroad(1),        1, MPI_INTEGER, rankdown, 1,  &
                       oldroad(nlocal+1), 1, MPI_INTEGER, rankup, 1,    &
                       MPI_COMM_WORLD, status, ierr);

     nmovelocal = updateroad(newroad, oldroad, nlocal)

! Globally sum the value

     call MPI_Allreduce(nmovelocal, nmove, 1, MPI_INTEGER, MPI_SUM, &
                        MPI_COMM_WORLD, ierr)

! Copy new to old array

     do i = 1, nlocal

        oldroad(i) = newroad(i)

     end do

     if (mod(iter, printfreq) == 0) then

        if (rank == 0) write(*,*) 'At iteration ', iter, &
             ' average velocity is ', float(nmove)/float(ncars)
     end if

  end do

  tstop = MPI_Wtime()

  deallocate(oldroad)
  deallocate(newroad)

  if (rank == 0) then

     write(*,*)
     write(*,*) 'Finished'
     write(*,*)
     write(*,*) 'Time taken was  ', tstop - tstart, ' seconds'
     write(*,*) 'Update rate was ', &
                 1.d-6*float(ncell)*float(maxiter)/(tstop-tstart), ' MCOPs'
     write(*,*)

  end if

  call MPI_Finalize(ierr)

end program traffic

File: code/C-MPI/traffic.c
#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

#include "uni.h"
#include "traffic.h"

int main(int argc, char **argv)
{
  // Set the size of the road

  long int ncell = 5120000;

  int *oldroad, *newroad;

  int i, iter, nmove, nmovelocal, ncars;
  int maxiter, printfreq;

  float density;

  double tstart, tstop;

  MPI_Status status;
  int rank, size, nlocal, rankup, rankdown;

  maxiter = 1000 /* 1.024e10/((double) ncell) */;
  printfreq = maxiter/10;

  // Set target density of cars

  density = 0.52;

  // Start MPI

  MPI_Init(NULL, NULL);

  // Find size and rank

  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  ncell *= size;
  nlocal = ncell/size;

  if (rank == 0)
    {
      printf("Running message-passing traffic model\n");
      printf("\nLength of road is %ld\n", ncell);
      printf("Number of iterations is %d \n", maxiter);
      printf("Target density of cars is %f \n", density);
      printf("Running on %d processes\n", size);
    }

  oldroad = (int *) malloc((nlocal+2)*sizeof(int));
  newroad = (int *) malloc((nlocal+2)*sizeof(int));

  for (i=1; i <= nlocal; i++)
    {
      oldroad[i] = 0;
      newroad[i] = 0;
    }

  if (rank == 0) {

      // Initialise road accordingly using random number generator

      printf("Initialising road ...\n");
  }

  // seed random number generator
  rinit(SEED);
  int ncars_local = 0;
  for (int i=0; i<=rank; i++) {
      ncars_local = initroad(&oldroad[1], nlocal, density);
  }
  MPI_Reduce(&ncars_local, &ncars, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

  if (rank == 0) {
      printf("...done\n");
      printf("Actual density is %f\n", (float) ncars / (float) ncell);
      printf("Scattering data ...\n");
    }


  if (rank == 0)
      {
          printf("... done\n\n");
      }

  // Compute neighbours

  rankup   = (rank + 1) % size;
  rankdown = (rank + size - 1) % size;

  MPI_Barrier(MPI_COMM_WORLD);

  tstart = MPI_Wtime();

  for (iter=1; iter<=maxiter; iter++)
    {

      // Implement halo swaps which now includes boundary conditions

      MPI_Sendrecv(&oldroad[nlocal], 1, MPI_INT, rankup, 1,
		   &oldroad[0],      1, MPI_INT, rankdown, 1,
		   MPI_COMM_WORLD, &status);

      MPI_Sendrecv(&oldroad[1],        1, MPI_INT, rankdown, 1,
		   &oldroad[nlocal+1], 1, MPI_INT, rankup, 1,
		   MPI_COMM_WORLD, &status);

      // Apply CA rules to all cells

      nmovelocal = updateroad(newroad, oldroad, nlocal);

      // Globally sum the value

      MPI_Allreduce(&nmovelocal, &nmove, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

      // Copy new to old array

      for (i=1; i<=nlocal; i++)
	{
	  oldroad[i] = newroad[i];
	}

      if (iter%printfreq == 0)
	{
	  if (rank == 0)
	    {
	      printf("At iteration %d average velocity is %f \n",
		     iter, (float) nmove / (float) ncars);
	    }
	}
    }

  MPI_Barrier(MPI_COMM_WORLD);

  tstop = MPI_Wtime();

  free(oldroad);
  free(newroad);

  if (rank == 0)
    {

      printf("\nFinished\n");
      printf("\nTime taken was  %f seconds\n", tstop-tstart);
      printf("Update rate was %f MCOPs\n\n", \
      1.e-6*((double) ncell)*((double) maxiter)/(tstop-tstart));
    }

  // Finish

  MPI_Finalize();
}
